pkg/controller/elasticache/user/setup.go

```diff
@@ -0,0 +1,185 @@
+package user
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	svcsdk "github.com/aws/aws-sdk-go/service/elasticache"
+	svcsdkapi "github.com/aws/aws-sdk-go/service/elasticache/elasticacheiface"
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/ratelimiter"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/pkg/errors"
+
+	svcapitypes "github.com/crossplane/provider-aws/apis/elasticache/v1alpha1"
+	awsclients "github.com/crossplane/provider-aws/pkg/clients"
+	"github.com/crossplane/provider-aws/pkg/clients/elasticache"
+)
+
+// SetupUser adds a controller that reconciles User.
+func SetupUser(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter, poll time.Duration) error {
+	name := managed.ControllerName(svcapitypes.UserGroupKind)
+	opts := []option{
+		func(e *external) {
+			c := &custom{client: e.client, kube: e.kube, external: e}
+			e.isUpToDate = c.isUpToDate
+			e.preCreate = c.preCreate
+			e.postCreate = postCreate
+			e.preObserve = preObserve
+			e.preDelete = preDelete
+			e.postObserve = postObserve
+			e.postDelete = postDelete
+			e.preUpdate = c.preUpdate
+			e.postUpdate = c.postUpdate
+			e.filterList = filterList
+		},
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewController(rl),
+		}).
+		For(&svcapitypes.User{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(svcapitypes.UserGroupVersionKind),
+			managed.WithInitializers(managed.NewDefaultProviderConfig(mgr.GetClient())),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), opts: opts}),
+			managed.WithPollInterval(poll),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type custom struct {
+	kube     client.Client
+	client   svcsdkapi.ElastiCacheAPI
+	external *external
+}
+
+func filterList(cr *svcapitypes.User, obj *svcsdk.DescribeUsersOutput) *svcsdk.DescribeUsersOutput {
+	userIdentifier := awsclients.String(meta.GetExternalName(cr))
+	resp := &svcsdk.DescribeUsersOutput{}
+	for _, user := range obj.Users {
+		if awsclients.StringValue(user.UserId) == awsclients.StringValue(userIdentifier) {
+			resp.Users = append(resp.Users, user)
+			break
+		}
+	}
+	return resp
+}
+
+func (e *custom) preCreate(ctx context.Context, cr *svcapitypes.User, obj *svcsdk.CreateUserInput) error {
+	obj.UserId = awsclients.String(cr.Name)
+	obj.UserName = awsclients.String(cr.Name)
```

Comment:
```suggestion
	obj.UserId = awsclients.String(meta.GetExternalName(cr))
	obj.UserName = awsclients.String(meta.GetExternalName(cr))
```

`cr.Name` should never be used since it may not be mutable by the user in composition.

Reply:
done